keps/sig-node/4680-add-resource-health-to-pod-status/README.md

```diff
@@ -0,0 +1,408 @@
+
+Today, the typical design is for jobs consuming a failing device to fail itself with the specific error code whenever possible. For the inference of long running workloads, k8s will keep restarting the workload without reallocating it on a different device. So container will be in crash loop backoff with limited information on why it is crashing.
+
+People develop strategies to deal with such situations. Exposing unhealthy devices in Pod Status will provide a generic way to understand that the failure is related to the unhealthy device and be able to respond to this properly.
+
+### Goals
+
+- Expose device health information (served by Device Plugin or DRA) in Pod Status and events.
+
+### Non-Goals
+
+- Expose any other device information beyond the health.
+- Expose CPU assignment of the pod by CPU manager or any other resources assignment by other managers.
+
+## Proposal
+
+### PodStatus.AllocatedResourcesStatus
+
+As part of the InPlacePodVerticalScaling KEP, the two new fields were introduced in Pod Status to reflect the currently allocated resources for the Pod:
+
+```
+	// AllocatedResources represents the compute resources allocated for this container by the
+	// node. Kubelet sets this value to Container.Resources.Requests upon successful pod admission
+	// and after successfully admitting desired pod resize.
+	// +featureGate=InPlacePodVerticalScaling
+	// +optional
+	AllocatedResources ResourceList `json:"allocatedResources,omitempty" protobuf:"bytes,10,rep,name=allocatedResources,casttype=ResourceList,castkey=ResourceName"`
+
+	// Resources represents the compute resource requests and limits that have been successfully
+	// enacted on the running container after it has been started or has been successfully resized.
+	// +featureGate=InPlacePodVerticalScaling
+	// +optional
+	Resources *ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,11,opt,name=resources"`
+```
+
+One field reflects the resource requests and limits and the other actual allocated resources.
+
+This structure will contain standard resources as well as extended resources. As noted in the comment: https://github.com/kubernetes/kubernetes/pull/124227#issuecomment-2130503713, it is only logical to also include the status of those allocated resources. 
+
+The proposal is to keep this structure as-is to simplify parsing of well-known ResourceList data type by various consumers. Typical scenario would be to compare if the AllocatedResources match the desired state.
+
+The proposal is to introduce an additional field:
+
+```
+// AllocatedResourcesStatus represents the status of various resources
+// allocated for this Pod.
+AllocatedResourcesStatus ResourcesStatus 
+
+type ResourcesStatus map[ResourceName]ResourceStatus
+
+type ResourceStatus struct {
+    // map of unique Resource ID to its status
+    Resources map[ResourceID] ResourceStatus
```

Comment:
how do you obtain ResourceID and map to a device?

Reply:
Furthermore, what does "resourceID" mean int eh context of DRA?

Reply:
Are you referring to device id here? https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/api.proto#L102-L106

Is that something that should be exposed to end users? e.g. if device plugin exposes some random GUID it would be hard to diagnose, it seems it would require device plugin to expose some ID name / ID that has some semantic meaning.

Reply:
For DRA, we will have `<driver name>/<node name>/<device name>` in 1.31 which has some meaning for certain devices. But not everything is using such an identifier.

The main identifier in the kubelet <-> DRA driver gRPC is the claim UID/name/namespace. A claim then may stand for several different devices. But surfacing that here is probably not useful.

Perhaps we can limit this KEP to those DRA devices which have a simple string identifier, like the one above? I'm still not sure how to properly define the format of `ResourceID` in a way that it can be parsed reliably, though.

Reply:
Starting with the device plugin sounds like a good option. I added a note to the ResourceID that it must not be compared between Pods and should only be unique for the lifetime of the Pod.

Reply:
The current thinking is that DRA will have a unique device ID string for all devices. The problem with `ResourceID` is that it could be an ID from different sub-systems and those sub-systems have different IDs.

So perhaps `ResourceID` needs to be defined as:
- `dra:<driver name>[/<pool name>]/<device name>`: such a device can be looked up in the information published by that DRA driver to learn more about it.
- `deviceplugin:<something>` - but what is that `<something>`?

What this KEP needs to define is what a consumer can do with `ResourceID`. If it's an opaque string, then what is the purpose of providing the status when no-one know what it is the status of?

Reply:
Thank you! Updated based on suggestion